### PR TITLE
when async operation fails with ERROR_OPERATION_ABORTED, we should ignore numBytes that can be != 0

### LINKED ISSUE #56894: FileStream_WriteAsync_AsyncWrites.WriteAsyncCancelledFile test failed
```

C:\h\w\AF2309CA\w\B67809CD\e>"C:\h\w\AF2309CA\p\dotnet.exe" exec --runtimeconfig System.IO.FileSystem.Tests.runtimeconfig.json --depsfile System.IO.FileSystem.Tests.deps.json xunit.console.dll System.IO.FileSystem.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing  
  Discovering: System.IO.FileSystem.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.IO.FileSystem.Tests (found 3576 of 3982 test cases)
  Starting:    System.IO.FileSystem.Tests (parallel test collections = on, max threads = 2)
    System.IO.Tests.DirectoryInfo_Create.RootPath_AppContainer [SKIP]
      Condition(s) not met: "IsInAppContainer"
    System.IO.Tests.Directory_CreateDirectory.RootPath_AppContainer [SKIP]
      Condition(s) not met: "IsInAppContainer"
    System.IO.Tests.FileInfo_GetSetTimes.CopyToMillisecondPresent_LowTempRes [SKIP]
      Condition(s) not met: "LowTemporalResolution"
    System.IO.Tests.FileInfo_GetSetTimes.MoveToMillisecondPresent_LowTempRes [SKIP]
      Condition(s) not met: "LowTemporalResolution"
    System.IO.Tests.FileInfo_GetSetTimes.CopyToNanosecondsPresent_LowTempRes [SKIP]
      Condition(s) not met: "LowTemporalResolution"
    System.IO.Tests.FileInfo_GetSetTimes.TimesIncludeMillisecondPart_LowTempRes [SKIP]
      Condition(s) not met: "LowTemporalResolution"
    System.IO.Tests.DirectoryInfo_GetSetTimes.TimesIncludeMillisecondPart_LowTempRes [SKIP]
      Condition(s) not met: "LowTemporalResolution"
    System.IO.Tests.FileStream_SafeFileHandle.ThrowWhenHandlePositionIsChanged_async [SKIP]
      Condition(s) not met: "IsNet5CompatFileStreamEnabled"
    System.IO.Tests.FileStream_SafeFileHandle.ThrowWhenHandlePositionIsChanged_sync [SKIP]
      Condition(s) not met: "IsNet5CompatFileStreamEnabled"
    System.IO.Tests.DirectoryInfo_ToString.ParentToString_Framework [SKIP]
      Condition(s) not met: "IsNetFramework"
    System.IO.Tests.File_GetSetTimes.TimesIncludeMillisecondPart_LowTempRes [SKIP]
      Condition(s) not met: "LowTemporalResolution"
    System.IO.Tests.Directory_GetSetTimes.TimesIncludeMillisecondPart_LowTempRes [SKIP]
      Condition(s) not met: "LowTemporalResolution"
    System.IO.Tests.FileStream_WriteAsync_AsyncWrites.WriteAsyncCancelledFile(bufferSize: 0, isAsync: True) [FAIL]
      Assert.Equal() Failure
      Expected: 0
      Actual:   189576032
      Stack Trace:
        /_/src/libraries/System.IO.FileSystem/tests/FileStream/WriteAsync.cs(128,0): at System.IO.Tests.FileStream_AsyncWrites.WriteAsyncCancelledFile(Int32 bufferSize, Boolean isAsync)
        --- End of stack trace from previous location ---
    System.IO.Tests.FileStream_WriteAsync_AsyncWrites.WriteAsyncCancelledFile(bufferSize: 4096, isAsync: True) [FAIL]
      Assert.Equal() Failure
      Expected: 0
      Actual:   192495168
      Stack Trace:
        /_/src/libraries/System.IO.FileSystem/tests/FileStream/WriteAsync.cs(128,0): at System.IO.Tests.FileStream_AsyncWrites.WriteAsyncCancelledFile(Int32 bufferSize, Boolean isAsync)
        --- End of stack trace from previous location ---
    System.IO.Tests.FileStream_ctor_options_as.WhenFileIsTooLargeTheErrorMessageContainsAllDetails [SKIP]
      Condition(s) not met: "IsFat32"
    System.IO.Tests.FileInfo_Open_options_as.WhenFileIsTooLargeTheErrorMessageContainsAllDetails [SKIP]
      Condition(s) not met: "IsFat32"
    System.IO.Tests.File_Open_str_options_as.WhenFileIsTooLargeTheErrorMessageContainsAllDetails [SKIP]
      Condition(s) not met: "IsFat32"
    System.IO.Tests.File_OpenHandle.WhenFileIsTooLargeTheErrorMessageContainsAllDetails [SKIP]
      Condition(s) not met: "IsFat32"
  Finished:    System.IO.FileSystem.Tests
=== TEST EXECUTION SUMMARY ===
   System.IO.FileSystem.Tests  Total: 7961, Errors: 0, Failed: 2, Skipped: 16, Time: 35.218s
----- end Thu 08/05/2021  7:51:22.22 ----- exit code 1 ----------------------------------------------------------
```

Configuration: `Libraries Test Run release coreclr windows x86 Debug`

Console log: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-56712-merge-c083d1b3869d4a13bc/System.IO.FileSystem.Tests/1/console.7aac5a3d.log?sv=2019-07-07&se=2021-08-25T07%3A40%3A08Z&sr=c&sp=rl&sig=OvyvEWqdM7XTBoPT9NzUEPfA21XoOfUo2%2BjKOeV6MUc%3D

Build: https://dev.azure.com/dnceng/public/_build/results?buildId=1277582&view=logs&j=9d8498d2-c5b7-54d8-6df7-a2ce7e14e68c&t=34b459a3-e64e-5ccc-0adc-77ebca2601a8&l=71

### PR DESCRIPTION
**Pull request by adamsitnik**
In #57212, I've validated that when async `WriteFile` operation fails with `ERROR_OPERATION_ABORTED`, the reported number of transferred bytes can be different than zero. Moreover, its value can be bigger than the number of bytes that we have requested to transfer when the operation was started, so we can assume that this value is always invalid in such case and just use zero instead.

```log
Process terminated. Assertion failed.
Callback returned 995 error and 1547079704 bytes
   at Microsoft.Win32.SafeHandles.SafeFileHandle.OverlappedValueTaskSource.Complete(UInt32 errorCode, UInt32 numBytes) in /_/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.OverlappedValueTaskSource.Windows.cs:line 198
   at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pNativeOverlapped) in /_/src/coreclr/System.Private.CoreLib/src/System/Threading
```

```log
System.IO.Tests.FileStream_WriteAsync_AsyncWrites.TriggerTheProblemAsync [FAIL]
  System.Exception : The error code was 995 and numBytes was 189216864
  Stack Trace:
    /_/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.OverlappedValueTaskSource.Windows.cs(101,0): at Microsoft.Win32.SafeHandles.SafeFileHandle.OverlappedValueTaskSource.GetResult(Int16 token)
    /_/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.OverlappedValueTaskSource.Windows.cs(96,0): at Microsoft.Win32.SafeHandles.SafeFileHandle.OverlappedValueTaskSource.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
    /_/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs(251,0): at System.Threading.Tasks.ValueTask.ValueTaskSourceAsTask.<>c.<.cctor>b__4_0(Object state)
    --- End of stack trace from previous location ---
    /_/src/libraries/System.IO.FileSystem/tests/FileStream/WriteAsync.cs(131,0): at System.IO.Tests.FileStream_AsyncWrites.WriteAsyncCancelledFile(Int32 bufferSize, FileOptions fileOptions)
    /_/src/libraries/System.IO.FileSystem/tests/FileStream/WriteAsync.cs(110,0): at System.IO.Tests.FileStream_AsyncWrites.TriggerTheProblemAsync()
    --- End of stack trace from previous location ---
``` 

fixes #56894

### CODE REVIEW

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.OverlappedValueTaskSource.Windows.cs [hunk 1]
```diff
@@ -195,34 +195,33 @@ private static void IOCallback(uint errorCode, uint numBytes, NativeOverlapped*
...
                     case Interop.Errors.ERROR_OPERATION_ABORTED:
+                        strategy?.OnIncompleteOperation(_bufferSize, 0); // don't use numBytes here, as it can be != 0 for this errorCode (#57212)
                         // Cancellation
                         CancellationToken ct = _cancellationRegistration.Token;
                         _source.SetException(ct.IsCancellationRequested ? new OperationCanceledException(ct) : new OperationCanceledException());
                         break;
 
                     default:
                         // Failure
+                        strategy?.OnIncompleteOperation(_bufferSize, 0);
```
Comment (carlossanlop) [inline src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.OverlappedValueTaskSource.Windows.cs:224]: Do we care to add a comment here explaining why 0 for the default case?

**Review by carlossanlop — APPROVED** [review]
LGTM. Just a small optional suggestion to add a comment.

File: src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.OverlappedValueTaskSource.Windows.cs [hunk 2]
```diff
@@ -195,34 +195,33 @@ private static void IOCallback(uint errorCode, uint numBytes, NativeOverlapped*
...
+                        if (_bufferSize != numBytes) // true only for incomplete operations
+                        {
+                            strategy?.OnIncompleteOperation(_bufferSize, (int)numBytes);
+                        }
                         // Success
                         _source.SetResult((int)numBytes);
                         break;
 
                     case Interop.Errors.ERROR_OPERATION_ABORTED:
+                        strategy?.OnIncompleteOperation(_bufferSize, 0); // don't use numBytes here, as it can be != 0 for this errorCode (#57212)
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.OverlappedValueTaskSource.Windows.cs:216]: My usual comment that #57212 can immediately be found by git blame here.

Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.OverlappedValueTaskSource.Windows.cs:216]: (but don't feel obliged to change it.)

Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.OverlappedValueTaskSource.Windows.cs:216]: PR #57212 was opened to test Adam's hypothesis, which he then closed when it was confirmed. Unless I'm missing something, git blame only shows merged PRs.

Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.OverlappedValueTaskSource.Windows.cs:216]: Ah true. Nevertheless, my personal preference would be to document whatever we learned right here (actually I think that's already done)

I know I'm a bit odd (although I think @stephentoub has the same general preference, we aren't consistent) and don't feel obliged to do anything here. We ought to reach collective consensus at some future point...

**Comment by ChayoteJarocho** [comment]
The code only affects Windows. The CI failure happened in an OSX unit test which @Jozkee fixed in #57235. This is good to merge.

**Comment by carlossanlop** [comment]
The code only affects Windows. The CI failure happened in an unrelated OSX unit test which @Jozkee fixed in #57235. This is good to merge.

**Pull request merged**